buddy_list: Fix incorrect user count in the right sidebar

arthurzengg opened the pull request:
My main changes are in one file: buddy_list.ts.

The issue occurs because of the following calculation:

1. `{N: total_human_subscribers_count - participant_count}`in `initialize_tooltips`
2. `const subscriber_section_user_count = total_human_subscribers_count - all_participant_ids.size;` in `update_section_header_count`
3. `user_count: get_formatted_sub_count(total_human_subscribers_count - all_participant_ids.size,)` in `render_section_headers`

When a conversation participant unsubscribes (e.g., user ID 6), but that user has previously posted something in the channel, they remain in all_participant_ids (i.e., {6}) because they are still considered a participant. However, since they have unsubscribed, they are no longer included in total_human_subscribers_count. As a result, the calculation produces an incorrect value, such as 0 - 1 = -1

### Solution Approach
To fix this issue, I followed the suggested approach of calculating subscribed_non_participant_count using the set-subtracts method. This avoids iterating over IDs directly, which could lead to performance issues.

1. Helper Function

I introduced the following helper function to compute the number of subscribed users who are not participants:
`function get_non_participant_subscribers_count`

The purpose of this helper function is to calculate the count once in get_render_data, and subscriber_section_user_count can be used in `render_section_headers`, and do not have to calculate again.

2. New Calculation for subscriber_section_user_count

Instead of relying on a subtraction approach that mixes different categories of users, I implemented a separate calculation:
`const non_participant_subscribers_count = get_non_participant_subscribers_count(
            this.render_data.current_sub,
            this.render_data.get_all_participant_ids,
        );`

This ensures that non_participant_subscribers_count is calculated independently without mixing the counts of unsubscribed users and participants.

Changes Applied in Three Places

1. initialize_tooltips – Updates the count when hovering over “THIS CHANNEL”.
![image](https://github.com/user-attachments/assets/a47db46a-5205-4687-bd41-ac3e28f844a2)
2. update_section_header_count – Updates the displayed number of users in “THIS CHANNEL”.
![image](https://github.com/user-attachments/assets/0c95af45-ffde-49ce-8b97-733e3c79a8a7)
3. render_section_headers – Ensures the correct count is displayed when the section is collapsed.
![image](https://github.com/user-attachments/assets/aaf6902d-4a59-406b-8550-183e06fe63a0)

Fixes: https://github.com/zulip/zulip/issues/33452

<details>
<summary>Self-review checklist</summary>
- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Harshbansal8705 commented:
Hi @arthurzengg, I see that you've just added a logic to show the count as `0` if it's less than `0`. But, I just wanted to know why does it show `-1` at all after the calculations? Have you tried to get the reason for this behaviour?

Harshbansal8705 commented:
Also, take a look at [this case](https://chat.zulip.org/#narrow/channel/9-issues/topic/incorrect.20user.20count.20in.20the.20right.20sidebar/near/2093216). I'm not sure, if your fix would work for that.

arthurzengg commented:
> Also, take a look at [this case](https://chat.zulip.org/#narrow/channel/9-issues/topic/incorrect.20user.20count.20in.20the.20right.20sidebar/near/2093216). I'm not sure, if your fix would work for that.

Hi Harshbansal, I provided a complete solution in my previous PR [#33491](https://github.com/zulip/zulip/pull/33491). Here’s what that PR addressed:
1. Adding a new helper function, get_unsubscribed_posters_count, that iterates over all conversation participants and counts those not subscribed to the current stream.
2. Updating the BuddyListRenderData structure to include the unsubscribed_posters_count.
3. Adjusting the logic in update_section_header_counts and the tooltip rendering in render_section_headers so that the final count now correctly adds the unsubscribed users. For example, the subscriber section now uses:
“total_human_subscribers_count – all_participant_ids.size + unsubscribed_posters_count”

This approach handles all scenarios. However, TimAbbott might have considered a different scope for the issue, which is why I closed that PR. If you think that solution covers the needed scenarios, I can reopen the PR. Thanks。

timabbott commented:
I think we do want to handle all scenarios; I just wasn't convinced that adding additional calculations, rather than adjusting existing ones, was the right way to do that. @evykassirer perhaps you can look at this.

arthurzengg commented:
My main changes are in one file: buddy_list.ts.

The issue occurs because of the following calculation: 
1. `{N: total_human_subscribers_count - participant_count}` in `initialize_tooltips`
2. `const subscriber_section_user_count = total_human_subscribers_count - all_participant_ids.size;` in `update_section_header_count`
3. `user_count: get_formatted_sub_count(total_human_subscribers_count - all_participant_ids.size,)` in `render_section_headers`


When a conversation participant unsubscribes (e.g., user ID 6), but that user has previously posted something in the channel, they remain in all_participant_ids (i.e., {6}) because they are still considered a participant. However, since they have unsubscribed, they are no longer included in total_human_subscribers_count. As a result, the calculation produces an incorrect value, such as 0 - 1 = -1

### Solution Approach
To fix this issue, I followed the suggested approach of calculating subscribed_non_participant_count using the set-subtracts method. This avoids iterating over IDs directly, which could lead to performance issues.

1. Helper Function

I introduced the following helper function to compute the number of subscribed users who are not participants:
`function get_subscribed_non_participant_count` 

The purpose of this helper function is to calculate the count once in get_render_data, and subscriber_section_user_count can be used in `render_section_headers`, and do not have to calculate again.

2. New Calculation for subscriber_section_user_count

Instead of relying on a subtraction approach that mixes different categories of users, I implemented a separate calculation:
`const all_subs = stream_id ? new Set(peer_data.get_subscribers(stream_id)) : new Set();`
`const subs_who_werent_participants = all_subs.difference(all_participant_ids);`
`const subscriber_section_user_count = subs_who_werent_participants.size;`

This ensures that subscriber_section_user_count is calculated independently without mixing the counts of unsubscribed users and participants.

Changes Applied in Three Places
1. initialize_tooltips – Updates the count when hovering over “THIS CHANNEL”.
<img width="408" alt="Screenshot 2025-02-21 at 00 51 47" src="https://github.com/user-attachments/assets/32e5f1a0-1387-49d1-b7ca-d8df25dac740" />

2. update_section_header_count – Updates the displayed number of users in “THIS CHANNEL”.
<img width="269" alt="Screenshot 2025-02-21 at 00 51 05" src="https://github.com/user-attachments/assets/4d927c92-28be-4fb6-80df-7f6134fc28b6" />

3. render_section_headers – Ensures the correct count is displayed when the section is collapsed.
<img width="221" alt="Screenshot 2025-02-21 at 00 50 19" src="https://github.com/user-attachments/assets/8303285e-4556-4a70-89ba-8d1cb87fd1e8" />

Harshbansal8705 commented on web/src/buddy_list.ts:
```diff
@@ -62,6 +62,17 @@ function get_total_human_subscriber_count(
+        return 0;
+    }
+    const all_subs = new Set(peer_data.get_subscribers(current_sub.stream_id));
+    return [...all_subs].filter((id) => !get_all_participant_ids().has(id)).length;
```
I think we decided to use the `difference()` method here, instead of iterating over the array?

arthurzengg replied:
I’ve now updated the implementation to use difference() as we discussed. Thanks for the feedback.

Harshbansal8705 reviewed (changes requested):
@arthurzengg Please also check the tests which have failed. 

Also, The commit message doesn't seem to be following [Commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html). You'll need to update that


I haven't checked the code logic yet, I'll do that tomorrow. Till then you could make these changes.

arthurzengg commented:
> @arthurzengg Please also check the tests which have failed.
> 
> Also, The commit message doesn't seem to be following [Commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html). You'll need to update that
> 
> I haven't checked the code logic yet, I'll do that tomorrow. Till then you could make these changes.

I’ve revised the commit message to follow the commit discipline. Regarding the failed tests, my latest code now passes all automated tests.

Harshbansal8705 commented:
Did you run `tools/lint` on your local. Commit message still doesn't follow the Commit discipline.

arthurzengg commented:
> Did you run `tools/lint` on your local? Commit message still doesn't follow the Commit discipline.
> Please read the documentation carefully.

Sorry for the inconvenience, just fixed.

Harshbansal8705 commented on web/src/buddy_list.ts:
```diff
@@ -393,10 +416,12 @@ export class BuddyList extends BuddyListConf {
+        const stream_id = this.render_data.current_sub?.stream_id;
+        const all_subs = stream_id ? new Set(peer_data.get_subscribers(stream_id)) : new Set();
+        const subs_who_werent_participants = all_subs.difference(all_participant_ids);
+        const subscriber_section_user_count = subs_who_werent_participants.size;
```
Can't we use `const subscriber_section_user_count = this.render_data.subscriber_section_user_count` here, instead of recalculating it? 

Also, why the variable named `subscriber_section_user_count`? I can't tell what does it signify by looking at it. Please consider changing variable name to something like `non_participant_subscribers_count`.

Harshbansal8705 reviewed (changes requested):
@arthurzengg I've suggested some changes. Please consider these. Otherwise, it looks good to me. 

@evykassirer if you want to take a look at the changes.

Harshbansal8705 commented on web/src/buddy_list.ts:
```diff
@@ -62,6 +62,17 @@ function get_total_human_subscriber_count(
     return 0;
 }
 
+function get_subscribed_non_participant_count(
```
I think the function could be named as `get_non_participant_subscriber_count` or `get_non_participant_subscribers_count`?

alya commented:
Thanks for the review, @Harshbansal8705 !

@evykassirer tagging you for the next round of review on this one.

arthurzengg commented:
@evykassirer Could you please take a look?

evykassirer commented on web/src/buddy_list.ts:
```diff
@@ -62,6 +62,19 @@ function get_total_human_subscriber_count(
+    if (!current_sub) {
+        return 0;
+    }
+    const all_subs = new Set(peer_data.get_subscribers(current_sub.stream_id));
```
Name this `all_subscribers`, to avoid confusion with `sub` often standing for `subscription`

evykassirer commented on web/src/buddy_list.ts:
```diff
@@ -393,13 +419,17 @@ export class BuddyList extends BuddyListConf {
 
     update_section_header_counts(): void {
-        const {total_human_subscribers_count, other_users_count} = this.render_data;
+        const {other_users_count} = this.render_data;
```
`non_participant_subscribers_count` should be in the `render_data` already right?

evykassirer reviewed (commented):
Looks good except for two comments!

arthurzengg commented:
@evykassirer I’ve updated my code based on your two comments. Thanks for your help. Feel free to review it again when you have a chance.

timabbott commented on web/src/buddy_list.ts:
```diff
@@ -224,7 +250,7 @@ export class BuddyList extends BuddyListConf {
-                                    {N: total_human_subscribers_count - participant_count},
+                                    {N: non_participant_subscribers_count},
                                 );
                             } else if (current_sub) {
```
This can be a follow-up, but I wonder if we can use `non_participant_subscribers_count` all the other places in this function that we use `total_human_subscribers_count`? After all, if `participant_count` is 0, these should be the same. 

It would be more efficient to not compute both if they will have the same value in the only places we use the other one.

evykassirer replied:
Yeah great idea, I'll add that to my list for after this and #33578 get merged :)

timabbott commented on web/src/buddy_list.ts:
```diff
@@ -62,6 +62,19 @@ function get_total_human_subscriber_count(
+    const bot_ids = people.get_bot_ids();
+    const bot_ids_set = new Set(bot_ids);
+    return all_subscribers.difference(get_all_participant_ids()).difference(bot_ids_set).size;
+}
```
I think this function counts deactivated users who are subscribers; `peer_data.get_subscriber_count` does not. But I think that's the empty set, so maybe it's OK.

evykassirer replied:
@arthurzengg can you confirm this is the empty set?

timabbott commented:
This feels fairly inefficient -- it adds a loop through every user in the organization, which can be 10Ks, to every buddy list render. And I don't see any reason we'd need to do that -- if we'd doing an adjustment to the math, I don't see why this couldn't be just replacing

`total_human_subscribers_count - all_participant_ids.size`

with

`total_human_subscribers_count` - `human_participant_ids.size`

which would involve the new calculation of computing human participants, which is likely to be a loop over 10 users, not 10Ks.

evykassirer commented:
Following up on Tim's comment:

The reason it's looping through every user is the call to `get_bot_ids`. So we should calculate this in a different more efficient way.

Tim is essentially suggesting something like this: `total_human_subscribers_count - subscribed_participant_ids.size` -- where you take the participants and calculate which of them are subscribed, which does do a loop but a pretty small one since most conversations have fairly few participants.

evykassirer commented:
I ended up needing to do this for another PR I'm working on, so feel free to copy the code I wrote.

I put this before `populate`:

```js
    subscribed_participant_ids(): number {
        const {current_sub, pm_ids_set, get_all_participant_ids} = this.render_data;
        const participant_ids_list = [...get_all_participant_ids()];
        return participant_ids_list.filter(user_id => {
            if (current_sub) {
                return peer_data.is_user_subscribed(current_sub.stream_id, user_id);
            } 
            return pm_ids_set.has(user_id);
        }).length;
    }
```

You could have it return a set instead if you like that better :)

arthurzengg commented:
Thanks for the update, @evykassirer and @timabbott. 
I agree that using total_human_subscribers_count - subscribed_participant_ids.size is a more straightforward approach for calculating get_non_participant_subscribers_count.

In `initialize_tooltips`, I updated `non_participant_subscribers_count` all the other places in this function that we use total_human_subscribers_count base on the suggestion.

 I've updated my code. Feel free to review again when you have a chance.

timabbott commented:
I don't quite understand why your new function isn't a loop over participants...

evykassirer commented on web/src/buddy_list.ts:
```diff
@@ -218,31 +234,13 @@ export class BuddyList extends BuddyListConf {
+                                        "{N, plural, one {# other subscriber} other {# other subscribers}}",
+                                },
+                                {N: non_participant_subscribers_count},
+                            );
```
Keep the last `else` block that says "participants". For DMs this is the people in the DM, and the formal participants section is hidden.

I can see how this is confusing, and I was pondering if we should change how it works -- but the users matching narrow section is for all people who can see the narrow, and the first section is for users who have sent messages in that narrow (which isn't necessarily all people in a DM conversation audience).

evykassirer commented on web/src/buddy_list.ts:
```diff
@@ -89,6 +101,7 @@ function get_render_data(): BuddyListRenderData {
     const other_users_count = people.get_active_human_count() - total_human_subscribers_count;
     const hide_headers = should_hide_headers(current_sub, pm_ids_set);
     const get_all_participant_ids = buddy_data.get_conversation_participants_callback();
+    const non_participant_subscribers_count = get_non_participant_subscribers_count(current_sub);
```
Unfortunately I don't think we can actually do this, which I didn't realize until yesterday, so sorry for misdirecting you. 

Reasoning: At the time `get_render_data` is called, the participants aren't loaded yet. This is why `get_all_participant_ids` is a function to be called later, instead of just an array of ids.

I'd recommend making a function `non_participant_subscribers_count` below `subscribed_participant_ids` that you call each time you need this value.

evykassirer commented:
> I don't quite understand why your new function isn't a loop over participants...

@timabbott are you referring to `subscribed_participant_ids` which is a loop (`filter`) over participants? Or `non_participant_subscribers_count` which needs to count subscribers with a set difference?

timabbott commented:
I was referring to `non_participant_subscribers_count`, which seems to still be doing a loop over everyone.

evykassirer commented:
Thanks for your work on this @arthurzengg! I've cleaned this up and opened #33776 with the continued work and closing this PR in favor of that one.
